src/builder/QuarkBuilder.sol

```diff
@@ -266,53 +266,93 @@ contract QuarkBuilder {
+        Actions.Action[] memory transferActions = findActionsOfType(actions, Actions.ACTION_TYPE_TRANSFER);
+
         uint256 plannedBridgeAmount = 0;
-        for (uint256 i = 0; i < actions.length; ++i) {
-            address tokenUsed;
-            uint256 amountUsed;
-            if (Strings.stringEqIgnoreCase(actions[i].actionType, Actions.ACTION_TYPE_TRANSFER)) {
-                Actions.TransferActionContext memory transferActionContext =
-                    abi.decode(actions[i].actionContext, (Actions.TransferActionContext));
-                tokenUsed = transferActionContext.token;
-                amountUsed = transferActionContext.amount;
-            } else if (Strings.stringEqIgnoreCase(actions[i].actionType, Actions.ACTION_TYPE_BRIDGE)) {
-                Actions.BridgeActionContext memory bridgeActionContext =
-                    abi.decode(actions[i].actionContext, (Actions.BridgeActionContext));
-                tokenUsed = bridgeActionContext.token;
-                amountUsed = bridgeActionContext.amount;
-                plannedBridgeAmount += amountUsed;
+        // Verify bridge actions is affordable, and update plannedBridgeAmount for verifying transfer actions
+        for (uint256 i = 0; i < bridgeActions.length; ++i) {
+            Actions.BridgeActionContext memory bridgeActionContext =
+                abi.decode(bridgeActions[i].actionContext, (Actions.BridgeActionContext));
+
+            if (bridgeActionContext.token == bridgeActions[i].paymentToken) {
+                // If the payment token is the transfer token and this is the target chain, we need to account for the transfer amount
+                uint256 paymentAssetBalanceOnChain = Accounts.sumBalances(
+                    Accounts.findAssetPositions(bridgeActionContext.token, bridgeActions[i].chainId, chainAccountsList)
+                );
+
+                // If its bridge step, check if user has enough balance to cover the bridge amount
+                if (paymentAssetBalanceOnChain < bridgeActions[i].paymentMaxCost + bridgeActionContext.amount) {
+                    revert MaxCostTooHigh();
+                }
             } else {
-                revert InvalidActionType();
+                // Just check payment token can cover the max cost
+                uint256 paymentAssetBalanceOnChain = Accounts.sumBalances(
+                    Accounts.findAssetPositions(
+                        bridgeActions[i].paymentToken, bridgeActions[i].chainId, chainAccountsList
+                    )
+                );
+
+                if (paymentAssetBalanceOnChain < bridgeActions[i].paymentMaxCost) {
+                    revert MaxCostTooHigh();
+                }
             }
 
-            if (tokenUsed == actions[i].paymentToken) {
+            plannedBridgeAmount += bridgeActionContext.amount;
+        }
+
+        // Verify transfer actions is affordable
+        uint256 accruedCost = 0;
```

Comment:
Hmm, I think this really needs to be "accrued cost on each chain" right? It wouldn't work if the transfers were on different chains. To reduce complexity for now, I think it's safe to assume there's only one transfer and remove this variable.

Reply:
Yeah was thinking to use mapping,  but we can't use mapping which complicates thing here, and ended up with one variable haha. Since we only have one transferIntent.chainId,  so I ended up with assuming the transfer(s) will be on one target chain.

Reply:
Now I have changed it to have capability to support multi-chain transfers.